interfaces/builtin/custom_device.go

```diff
@@ -0,0 +1,417 @@
+				}
+			}
+		case "attributes", "environment":
+			err = iface.validateUDevValueMap(value)
+		default:
+			err = errors.New(`unknown key`)
+		}
+
+		if err != nil {
+			return fmt.Errorf(`custom-device "udev-tagging" invalid %q tag: %v`, key, err)
+		}
+	}
+
+	if !hasKernelTag {
+		return errors.New(`custom-device udev tagging rule missing mandatory "kernel" key`)
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) Name() string {
+	return "custom-device"
+}
+
+func (iface *customDeviceInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              customDeviceSummary,
+		BaseDeclarationSlots: customDeviceBaseDeclarationSlots,
+	}
+}
+
+func (iface *customDeviceInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	if slot.Attrs == nil {
+		slot.Attrs = make(map[string]interface{})
+	}
+	customDeviceAttr, isSet := slot.Attrs["custom-device"]
+	customDevice, ok := customDeviceAttr.(string)
+	if isSet && !ok {
+		return fmt.Errorf(`custom-device "custom-device" attribute must be a string, not %v`,
+			customDeviceAttr)
+	}
+	if customDevice == "" {
+		// custom-device defaults to "slot" name if unspecified
+		slot.Attrs["custom-device"] = slot.Name
+	}
+
+	var devices []string
+	err := slot.Attr("devices", &devices)
+	if err != nil && !errors.Is(err, snap.AttributeNotFoundError{}) {
+		return err
+	}
+	for _, device := range devices {
+		if err := iface.validateDevice(device, "devices"); err != nil {
+			return err
+		}
+	}
+
+	var readDevices []string
+	err = slot.Attr("read-devices", &readDevices)
+	if err != nil && !errors.Is(err, snap.AttributeNotFoundError{}) {
+		return err
+	}
+	for _, device := range readDevices {
+		if err := iface.validateDevice(device, "read-devices"); err != nil {
+			return err
+		}
+		if strutil.ListContains(devices, device) {
+			return fmt.Errorf(`cannot specify path %q both in "devices" and "read-devices" attributes`, device)
+		}
+	}
+
+	allDevices := devices
+	allDevices = append(allDevices, readDevices...)
+
+	// validate files
+	var filesMap map[string][]string
+	err = slot.Attr("files", &filesMap)
+	if err != nil && !errors.Is(err, snap.AttributeNotFoundError{}) {
+		return err
+	}
+	for key, val := range filesMap {
+		switch key {
+		case "read":
+			if err := iface.validatePaths("read", val); err != nil {
+				return err
+			}
+		case "write":
+			if err := iface.validatePaths("write", val); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf(`cannot specify %q in "files" section, only "read" and "write" allowed`, key)
+		}
+	}
+
+	var udevTaggingRules []map[string]interface{}
+	err = slot.Attr("udev-tagging", &udevTaggingRules)
+	if err != nil && !errors.Is(err, snap.AttributeNotFoundError{}) {
+		return err
+	}
+	for _, udevTaggingRule := range udevTaggingRules {
+		if err := iface.validateUDevTaggingRule(udevTaggingRule, allDevices); err != nil {
+			return err
+		}
+	}
+
```

Comment:
should this also check:
```
if len(allDevices) == 0 && len(filesMap) == 0 {
    return fmt.Errorf("cannot use custom-device slot without any files or devices")
}
```

Reply:
this makes sense. Should it be even more strict though? does it make sense to have only files, isn't that like system-files? shouldn't we always have at leas one devices or read-devices?

Reply:
I think it's ok to have at least one item here, be it a file or a (read-)device. I can image a weird 3rd party driver that exposes things via some random /proc/ location only, but without any device nodes.

On a related note, some devices may be exposed under /dev but not have an entry in udev (eg. /dev/nvidia* from nvidia drivers used to be like this). AFAIU this will still work with custom-devices, but we'll still produce a udev match rule form it (although unused really).

Reply:
I implemented Maciej's suggestion, and added a unit test for it.